### PR TITLE
Posts bundle

### PR DESCRIPTION
**Pull request by valiknet18**

### CODE REVIEW

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    public function addPostAction(Request $request)
+    {
+        if($request->isMethod('POST')){
+            $em = $this->getDoctrine()->getManager();
+
+            $post = new Post();
+
+            $rq = $request->request;
+
+            $post->setTitle($rq->get('title'))
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: use 

``` php
$request->request->get('title');
```

for POST variables

File: src/Valiknet/Blog/PostsBundle/Entity/Comment.php~ [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+namespace Valiknet\Blog\PostsBundle\Entity;
+
+use Gedmo\Mapping\Annotation as Gedmo;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * @ORM\Entity
+ * @ORM\Table(name="comment")
+ */
+class Comment {
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Comment.php~]:
``` php
class Comment
{
```

File: src/Valiknet/Blog/PostsBundle/Entity/Comment.php~ [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * @ORM\Column(type="string", length=255)
+     */
+    protected $text;
+
+    /**
+     * @Gedmo\Timestampable(on="create")
+     * @ORM\Column(type="datetime")
+     */
+    protected $created_at;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Comment.php~]:
``` php
protected $createdAt;
```

File: src/Valiknet/Blog/PostsBundle/Entity/Post.php [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    /**
+     * @ORM\Column(type="string", length=200)
+     */
+    protected $author;
+
+    /**
+     * @Gedmo\Timestampable(on="create")
+     * @ORM\Column(type="datetime", name="createdAt")
+     */
+    protected $created_at;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Post.php]:
``` php
protected $createdAt;
```

File: src/Valiknet/Blog/PostsBundle/Entity/Post.php [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * @Gedmo\Timestampable(on="update")
+     * @ORM\Column(type="datetime", nullable=true, name="updatedAt")
+     */
+    protected $updated_at;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Post.php]: the same. use PSR-1 for naming convention

File: src/Valiknet/Blog/PostsBundle/Entity/Post.php [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * @ORM\Column(type="datetime", nullable=true, name="deletedAt")
+     */
+    protected $deleted_at;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Post.php]: use PSR-1 for naming convention

File: src/Valiknet/Blog/PostsBundle/Entity/Post.php [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * @Gedmo\Slug(fields={"title"})
+     * @ORM\Column(type="string", length=128, unique=true, name="slug")
+     */
+    protected $slug_post;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Post.php]: use PSR-1 for naming convention

File: src/Valiknet/Blog/PostsBundle/Entity/Post.php~ [hunk 8]
```diff
@@ -0,0 +1,69 @@
+<?php
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Post.php~]: remove this file and add 

```
*~
```

to your .gitignore

File: src/Valiknet/Blog/PostsBundle/Entity/Tag.php [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+     * @ORM\Column(type="integer")
+     * @ORM\Id
+     * @ORM\GeneratedValue(strategy="AUTO")
+     */
+    protected $id;
+
+    /**
+     * @ORM\Column(type="string", length=100, name="hashTag")
+     */
+    protected $hash_tag;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Tag.php]: use PSR-1 for naming convention

File: src/Valiknet/Blog/PostsBundle/Entity/Tag.php~ [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+     * @ORM\Column(type="integer")
+     * @ORM\Id
+     * @ORM\GeneratedValue(strategy="AUTO")
+     */
+    protected $id;
+
+    /**
+     * @ORM\Column(type="string", length=100, name="hashTag")
+     */
+    protected $hash_tag;
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Tag.php~]: use PSR-1 for naming convention

File: src/Valiknet/Blog/PostsBundle/Controller/CommentController.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+ */
+class CommentController extends Controller
+{
+    /**
+     * @Route("/add/{id}", name="comment_add")
+     * @Method({"POST"})
+     */
+    public function createCommentAction($id, Request $request)
+    {
+        if($request->isMethod('POST')){
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/CommentController.php]: You have annotation @Method and not needed to check it again

Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/CommentController.php]: also you miss spaces after "if" and before "{"

File: src/Valiknet/Blog/PostsBundle/Controller/CommentController.php [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+                json_encode(
+                    array(
+                        "code" => 200,
+                        "data" => $comments
+                    )
+                )
+            );
+        }
+
+        return new Response(
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/CommentController.php:42]: What the use case of this response?

File: src/Valiknet/Blog/PostsBundle/Controller/CommentController.php [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    /**
+     * @Route("/last", name="comment_last")
+     * @Method({"GET"})
+     * @Template("ValiknetBlogPostsBundle:Comment:last.html.twig")
+     */
+    public function getLastCommentsAction()
+    {
+        $comments = $this->getDoctrine()->getRepository('ValiknetBlogPostsBundle:Comment')->findBy([], ['id' => 'DESC'], 20);
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/CommentController.php]: what the 20? It's a "magic number", you need specify constant, or, better use parameters from url like

```
?limit=20&page=3
```

File: src/Valiknet/Blog/PostsBundle/Controller/CommentController.php [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                array(
+                    "code" => 404
+                )
+            ),
+            404
+        );
+    }
+
+    /**
+     * @Route("/last", name="comment_last")
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/CommentController.php]: what this resource is represent?

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route as Route;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method as Method;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template as Template;
+use Doctrine\ORM\Query\ResultSetMapping;
+use Valiknet\Blog\PostsBundle\Entity\Post;
+
+
+class PostController extends Controller
+{
+
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: extra line

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route as Route;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method as Method;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template as Template;
+use Doctrine\ORM\Query\ResultSetMapping;
+use Valiknet\Blog\PostsBundle\Entity\Post;
+
+
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: extra line

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+class PostController extends Controller
+{
+
+    /**
+     * @Route("/", name="blog_home")
+     * @Method({"GET"})
+     * @Template("ValiknetBlogPostsBundle:Post:index.html.twig")
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: use convention for template naming, no need to specify template name

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template as Template;
+use Doctrine\ORM\Query\ResultSetMapping;
+use Valiknet\Blog\PostsBundle\Entity\Post;
+
+
+class PostController extends Controller
+{
+
+    /**
+     * @Route("/", name="blog_home")
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php:17]: use convention for routing naming, no need to specify routing name

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+     * @Method({"GET"})
+     * @Template("ValiknetBlogPostsBundle:Post:index.html.twig")
+     */
+    public function indexAction()
+    {
+        $em = $this->getDoctrine()->getManager();
+
+        $posts = $em->getRepository('ValiknetBlogPostsBundle:Post')->findBy([], ['id' => 'DESC']);
+
+        $tags = $em->getRepository('ValiknetBlogPostsBundle:Tag')
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: move this into repository class

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+                                    ->createQueryBuilder('t')
+                                    ->groupBy('t.hashTag')
+                                    ->setMaxResults(10)
+                                    ->getQuery()
+                                    ->getResult();
+
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: extra lines

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+            return $this->redirect($this->get('router')->generate('blog_home'));
+        }
+
+        return array(
+            "tags" => $this->getDoctrine()->getRepository('ValiknetBlogPostsBundle:Tag')->findAll()
+        );
+    }
+
+    /**
+     * @Route("/post/view/{slug}", name="view_post")
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: no need to specify "view" in url - GET method is a view method

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        $post = $this->getDoctrine()->getRepository('ValiknetBlogPostsBundle:Post')->findOneBy(['slugPost' => $slug]);
+
+        return array(
+            "post" => $post
+        );
+    }
+
+    /**
+     * @Route("/post/edit/{slug}", name="edit_post")
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: for rest urls use "/posts/{slug}/edit"

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+     * @Template("ValiknetBlogPostsBundle:Post:edit.html.twig")
+     */
+     public function editPostAction($slug, Request $request)
+     {
+         if($request->isMethod('POST')){
+             $em = $this->getDoctrine()->getManager();
+
+             $post = $this->getDoctrine()->getRepository('ValiknetBlogPostsBundle:Post')->findOneBySlugPost($slug);
+
+             foreach($post->getTag() as $key=>$value){
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: fix code style please

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+         $tags = $this->getDoctrine()->getRepository('ValiknetBlogPostsBundle:Tag')->findAll();
+
+         return array(
+             "post" => $post,
+             "tags" => $tags
+         );
+     }
+
+    /**
+     * @Route("/post/delete/{id}", name="delete_post")
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: be consistent, if you view and edit by slug, so you need to delete by slug also

File: src/Valiknet/Blog/PostsBundle/Controller/PostController.php [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+
+        foreach($post->getTag() as $value){
+            $value->removePost($post);
+            $post->removeTag($value);
+        }
+
+        $em->remove($post);
+        $em->flush();
+
+        return new Response(
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/PostController.php]: use Symfony\Component\HttpFoundation\JsonResponse

File: src/Valiknet/Blog/PostsBundle/Controller/TagController.php [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * @Route("/last", name="tag_last_page")
+     * @Method({"GET"})
+     * @Template("ValiknetBlogPostsBundle:Tag:last.html.twig")
+     */
+    public function getLastTags()
+    {
+        $tags = $this->getDoctrine()->getRepository('ValiknetBlogPostsBundle:Tag')
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Controller/TagController.php]: move this to Repository class

File: src/Valiknet/Blog/PostsBundle/Entity/Post.php [hunk 27]
```diff
@@ -0,0 +1,313 @@
...
+    /**
+     * Get tag
+     *
+     * @return \Doctrine\Common\Collections\Collection 
+     */
+    public function getTag()
+    {
+        return $this->tag;
+    }
+
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Entity/Post.php]: extra line

File: src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+{% endblock %}
+
+{% block content %}
+    <div>
+        <ul>
+            {% for tag in tags %}
+                <li>
+                    <a href='{{ path('tag_page', { "slug" : tag.hashSlug }) }}'>{{ tag.hashTag }}</a>
+                </li>
+            {% else %}
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig:18]: oops, there must be error

File: src/Valiknet/Blog/PostsBundle/ValiknetBlogPostsBundle.php [hunk 29]
```diff
@@ -6,4 +6,18 @@
 
 class ValiknetBlogPostsBundle extends Bundle
 {
+    public function boot()
+    {
+        $doctrine = $this->container->get('doctrine');
```
Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/ValiknetBlogPostsBundle.php:11]: I think this can be configured.

**Comment by spolischook** [comment]
fix code style according to PSR-1 and PSR-2

File: src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig (hunk 28 above)
Comment (valiknet18) [inline src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig:18]: why?

Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig:18]: where your {% if %} ?

Comment (spolischook) [inline src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig:18]: oh, I see, a don't know that loop have else statement O_O

Comment (valiknet18) [inline src/Valiknet/Blog/PostsBundle/Resources/views/Tag/last.html.twig:18]: for/else statement very useful, not need write many if/else statement

File: src/Valiknet/Blog/PostsBundle/DataFixtures/ORM/LoadPostData.php [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+{
+    public function load(ObjectManager $manager)
+    {
+        $post = new Post();
+        $post->setTitle('Перший пост');
+        $post->setText('Lorem Ipsum - це текст-"риба", що використовується в друкарстві та дизайні. Lorem Ipsum є, фактично, стандартною "рибою" аж з XVI сторіччя, коли невідомий друкар взяв шрифтову гранку та склав на ній підбірку зразків шрифтів. "Риба" не тільки успішно пережила п\'ять століть, але й прижилася в електронному верстуванні, залишаючись по суті незмінною. Вона популяризувалась в 60-их роках минулого сторіччя завдяки виданню зразків шрифтів Letraset, які містили уривки з Lorem Ipsum, і вдруге - нещодавно завдяки програмам комп\'ютерного верстування на кшталт Aldus Pagemaker, які використовували різні версії Lorem Ipsum.');
+        $post->setAuthor('Гриневич В. О');
+        $post->setCreatedAt(new \DateTime());
+
+        $tag1 = new Tag();
```
Comment (piskovoy-dmitrij) [inline src/Valiknet/Blog/PostsBundle/DataFixtures/ORM/LoadPostData.php:19]: Если у тебя фикстуры для нескольких сущностей, то разделях их на отдельные файлы 
http://symfony.com/doc/current/bundles/DoctrineFixturesBundle/index.html#sharing-objects-between-fixtures

**Pull request merged**